Non-validated HttpHeaders enumeration

alnikola opened the pull request:
New `HttpHeaders.NonValidated` property returns a `HttpHeadersNonValidated` instance enumerating headers' raw values without any parsing or validation. It does **not change existing parsing logic**, but introduces a new `OriginalRawValue` property storing the raw header value. The old `RawValue` gets reset after each parsing, but `OriginalRawValue` preserves its value. If `RawValue` doesn't exist at all (in example, this happens when a value set through strongly-typed accessors), `HttpHeadersNonValidated` does call a parser to convert `ParsedValue` to a string representation the same way as it's done in the default validating enumeration logic.

@geoffkizer @karelz  This PR looks **totally safe** from the perspective of the existing parsing logic.

Fixes #35126

P.S. I will add a few more unit tests here.

alnikola committed:
IsParsed flag replaces RawValue's nullability checks

alnikola committed:
NonValidated uses OriginalRawValue as non-validating enumeration data source.

ManickaP commented on src/libraries/System.Net.Http/ref/System.Net.Http.cs:
```diff
@@ -505,6 +506,37 @@ public abstract partial class HttpHeaders : System.Collections.Generic.IEnumerab
+            void System.Collections.IEnumerator.Reset() { throw null; }
+        }
+    }
+    public readonly struct HeaderStringValues : System.Collections.Generic.IEnumerable<string>
```
Is this generated via [ref assembly generator](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/updating-ref-source.md)?
Is this placed where the generator placed it?

Because if not, it will cause plethora changes once somebody else uses the tool.

alnikola replied:
Oh, I missed it and added it manually. Will generate now. Thanks!

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+            ? new Enumerator(_headerStore)
+            : ((IEnumerable<KeyValuePair<string, HeaderStringValues>>)Array.Empty<KeyValuePair<string, HeaderStringValues>>()).GetEnumerator();
+
+        IEnumerator<KeyValuePair<string, HeaderStringValues>> IEnumerable<KeyValuePair<string, HeaderStringValues>>.GetEnumerator() => GetEnumerator();
```
Why do we need this? There's already public `GetEnumerator`. Doesn't it match the `IEnumerable<KeyValuePair<string, HeaderStringValues>>` contract?

ManickaP replied:
API proposal say that it should be returning `Enumerator`, but that's probably not possible.
If we're not exposing the `Enumerator`, couldn't it be private then?

alnikola replied:
It's an explicit interface implementation. But the above `GetEnumerator()` can return `Enumerator`. Will try to fix it.

alnikola replied:
Ah. No, it cannot, because it does a fast shortcut for an empty `_headerStore`.

geoffkizer replied:
We want the public GetEnumerator to return the Enumerator struct so that it doesn't get boxed.

geoffkizer replied:
This still needs to get fixed to return Enumerator here so we avoid boxing.

alnikola replied:
The above GetEnumerator definition is fixed. This line cannot be changed because it's part of interface.
```C#
public Enumerator GetEnumerator() => new Enumerator(_headerStore);
```

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+            _headerStore = headerStore;
+        }
+
+        public IEnumerator<KeyValuePair<string, HeaderStringValues>> GetEnumerator() => _headerStore != null
```
Couldn't the `Enumerator` itself handle the `_headerStore` being `null` instead of the `Array.Empty.GetEnumerator()` here? Our enumerator is a struct so it might even be cheaper.

alnikola replied:
This mirrors the current implementation in HttpHeaders.GetEnumerator.

stephentoub replied:
> This mirrors the current implementation in HttpHeaders.GetEnumerator.

But the HttpHeaders API returns the interface.  This needs to be strongly-typed as returning the struct.  That changes things.

alnikola replied:
Done.

alnikola committed:
Unnecessary nullability specification removed

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
```
Just a personal styling choice, but having the public implementation of an interface getter enhanced with setter seems a bit confusing.
Wouldn't it be a slightly readable to back it up with private `_current` field and leave the property to contain only getter and thus match the interface?

alnikola replied:
Personally, I'd prefer to have less code when it's possible and keep the current version. Adding a private doesn't change the interface in any way.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+            void IEnumerator.Reset()
+            {
+                _headerStore.Reset();
```
Shouldn't this reset `Current` to `default` as well?

alnikola replied:
Fixed.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+        private readonly IEnumerable<string?>? _values;
+        private readonly string? _value;
+
+        public HeaderStringValues(string value)
```
This shouldn't be public according to the API.

alnikola replied:
Hmm, as far as I see it, [the approved version](https://github.com/dotnet/runtime/issues/35126#issuecomment-670095983) states this should be `public`.

```C#
    public readonly struct HeaderStringValues : IEnumerable<string>
    {
        public HeaderStringValues(string value);
        public HeaderStringValues(IEnumerable<string> value);
        public Enumerator GetEnumerator();
        public readonly struct Enumerator : IEnumerator<string>
        {
            public bool MoveNext();
            public string Current { get; }
            public void Dispose();
            ... // explicitly implemented interface members
        }
    }
```

ManickaP replied:
Sorry, my bad, I wasn't looking properly at the API.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+            _values = null;
+        }
+
+        public HeaderStringValues(IEnumerable<string?> value)
```
This shouldn't be public according to the API.

alnikola replied:
Same as above.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+            private readonly string? _value;
+            private bool _completed;
+
+            public Enumerator(IEnumerable<string?>? values, string? value)
```
This shouldn't be public according to the API.

alnikola replied:
Same as above.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+        public Enumerator GetEnumerator() => new Enumerator(_values, _value);
+
+        IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumerator();
```
Do we need the explicit implementation?

alnikola replied:
It's required to implement all `IEnumerable<string>` members. The above `public Enumerator GetEnumerator()` is needed to support "duck typing" and avoid boxing, but it's not part of the `IEnumerable<string>` interface.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+                return true;
+            }
+
+            public string Current { get; private set; }
```
The same thing with the setter on interface implementation as above.

alnikola replied:
Discussed above.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
+            void IEnumerator.Reset()
+            {
+                _values?.Reset();
+                _completed = false;
```
Shouldn't this clear `Current` as well.

alnikola committed:
- IEnumerator.Reset resets Current
- ref is regenerated

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -759,8 +762,10 @@ private bool ParseRawHeaderValues(HeaderDescriptor descriptor, HeaderStoreItemIn
                     // At this point all values are either in info.ParsedValue, info.InvalidValue, or were removed since they
-                    // contain invalid newline chars. Reset RawValue.
+                    // contain invalid newline chars. Reset RawValue, but preserve OriginalRawValue to keep data for non-validating enumeration.
+                    object? originalRawValue = info.OriginalRawValue;
```
Why is this retrieving `info.OriginalRawValue` and then just setting it back to the same value?

geoffkizer replied:
And similarly below

alnikola replied:
It's a leftover from an initial version where `RawValue` setter also updated `OriginalRawValue`. I forgot to remove it, will do now.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1316,6 +1335,7 @@ internal class HeaderStoreItemInfo
             internal object? RawValue { get; set; }
             internal object? InvalidValue { get; set; }
             internal object? ParsedValue { get; set; }
+            internal object? OriginalRawValue { get; set; }
```
This increases the size of every HeaderStoreItemInfo by 20%. Why is that ok?

alnikola replied:
I realize that, and initially tried to remove `RawValue` resets and keep track everything there, but it proved to be extremely complex and dangerous because it requires changing the parsing logic in multiple places.

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -326,7 +328,7 @@ private string GetHeaderString(HeaderDescriptor descriptor, object info, object?
                     // and all values in it will be HeaderStoreItemInfo.
                     Debug.Assert(!_forceHeaderStoreItems);
-                    _headerStore[descriptor] = info = new HeaderStoreItemInfo() { RawValue = value };
+                    _headerStore[descriptor] = info = new HeaderStoreItemInfo() { RawValue = value, OriginalRawValue = value };
```
It would be nice to assert here that value is a string. I know that's not in the existing code, but it seems like it should be added.

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -570,7 +573,7 @@ internal virtual void AddHeaders(HttpHeaders sourceHeaders)
                     {
                         Debug.Assert(sourceValue is string);
-                        _headerStore.Add(header.Key, _forceHeaderStoreItems ? new HeaderStoreItemInfo { RawValue = sourceValue } : sourceValue);
+                        _headerStore.Add(header.Key, _forceHeaderStoreItems ? new HeaderStoreItemInfo { RawValue = sourceValue, OriginalRawValue = sourceValue } : sourceValue);
```
I don't understand why the existing code is not forcing parsing in this case. We force parsing above, so why not here? Seems inconsistent.

alnikola replied:
Sorry, I don't know right now. It would take more time to investigate. Anyways, this PR doesn't change this logic.

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -942,6 +947,9 @@ private static void AddValue(HeaderStoreItemInfo info, object? value, StoreLocat
                     currentStoreValue = info.RawValue;
                     AddValueToStoreValue<string>(value, ref currentStoreValue);
                     info.RawValue = currentStoreValue;
+                    currentStoreValue = info.OriginalRawValue;
```
It seems awkward and nonobvious to set both RawValue and OriginalRawValue here. Why not just add StoreLocation.OriginalRaw and make this explicit?

alnikola replied:
My goal is to minimize changes to the existing code, so I tried updated as few places as possible. If I implemented a new `StoreLocation.OriginalRaw` I would have to add `AddValue(StoreLocation.OriginalRaw)` calls to all places where  `AddValue(StoreLocation.Raw)` is called.

stephentoub replied:
> I would have to add AddValue(StoreLocation.OriginalRaw) calls to all places where AddValue(StoreLocation.Raw) is called.

How many are there?  Three?

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1024,6 +1032,12 @@ private void ParseAndAddValue(HeaderDescriptor descriptor, HeaderStoreItemInfo i
+            // We can get here only through Add methods which treat `value` as already parsed and do not add it into RawValue,
+            // so we have to manually add it into OriginalRawValue to return from non-validating enumeration.
+            object? originalRawValue = info.OriginalRawValue;
+            AddValueToStoreValue<string>(value ?? string.Empty, ref originalRawValue);
```
This is going to add overhead on every header value set by the user. I think it would be better to avoid that if it's not needed.

alnikola replied:
Could you please clarify how we can avoid it? We need to track all changes made by all `Add()` calls which do not changed `RawValue`.

geoffkizer replied:
Just clear out the OriginalRawValue. That's what you are doing for set and remove, right? Why can't you do the same thing for Add?

alnikola replied:
If we did that, we would lose the original raw value. In example, test [NonValidated_AddAndRemoveSingleValueHeader_EnumeratorReturnsNoValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1785) would fail because instead of the expected "raw1", it would get "parsed1". So, I think we need to keep it around.

geoffkizer replied:
Why do we care about keeping the original raw value in this case? If the user modifies the header through the API, then I think it's fine to discard the original raw value. In a lot of cases (like set and remove), we need to do discard it anyway, because it's not correct anymore. In this particular case, since they are adding the value as a string, I guess you could make the case that we can add another original raw value that corresponds to the value that they added. But it seems inconsistent, and I'm not sure what value it gets us. Seems better just to say, if you modify stuff via parsed header API (i.e. anything except TryAddWithoutValidation), you lose your original raw values.

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,158 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+            public Enumerator(IEnumerable<string>? values, string? value)
+            {
+                _values = values?.GetEnumerator();
```
I don't think this should be public.

Also, seems like it would be better to have two constructors here, one for single value, one for multi-value. That way the arguments don't need to be nullable.

geoffkizer commented:
There's a big comment at the top of the file describing the interaction between Raw/Parsed/etc values and how parsing works. Since we are making changes to this, we should update the comments to reflect the changes.

stephentoub commented:
@geoffkizer can make a call here, but I don't think to yes is the right aooroach. This is making existing code more expensive by increasing the size of the header object. I previously made a change that will lazy-allocate them only when needed, but any code that's enumerating the headers using the existing API or accessing the strongly-typed APIs will end up paying for this.

geoffkizer commented:
If I understand this PR correctly, here's the current approach (please correct me if I'm wrong):

(1) Every time a RawValue is added, we also add the same value as an OriginalRawValue.
(2) On parse, we parse the RawValue and set the ParsedValues, then clear the RawValue (as today), but we leave the existing OriginalRawValue in place
(3) On add of a strongly-typed value to ParsedValues, we also add a new OriginalRawValue.

Question: What happens if a user modifies an existing ParsedValue? E.g. update a single-valued header, or remove an item from a multi-valued header. Seems like we are not handling this case.

Beyond that, I have two concerns with the current implementation:
(1) It increases the size of the header store by adding OriginalRawValues (as @stephentoub mentioned above)
(2) It adds overhead in the common case of setting strongly typed values, because we need to also store an OriginalRawValue whenever this happens (as I mentioned above)

I'd suggest the following changes:

First, I don't think we actually need OriginalRawValue. We can just keep RawValue around after parsing is done. We can know the RawValues have been parsed by the presence of Parsed and/or Invalid values and not re-parse the RawValues again once it's parsed.

Second, I think we can avoid the overhead of keeping the RawValue in sync with any changes the user makes to ParsedValue by simply clearing the RawValue if they make changes. We can then generate a new RawValue if/when they enumerate NonValidatedHeaders. This also fixes the update issue I mentioned above. 

In short, at a given time, a header is in one of these states:
(1) Just parsed values. E.g. user adds parsed values. We can generate a raw value from these when needed (we do this today anyway when enumerating).
(2) Just raw values. E.g. user adds raw values via TryAddWithoutValidation. If the user does something to cause parsing, then we parse and end up in state (3) below.
(3) Both parsed values and raw values. In this case, the parsed values and raw values are always guaranteed to be in sync. If the user modifies either, we discard the other and end up in state (1) or (2).

Thoughts?

alnikola commented:
> First, I don't think we actually need OriginalRawValue. We can just keep RawValue around after parsing is done. We can know the RawValues have been parsed by the presence of Parsed and/or Invalid values and not re-parse the RawValues again once it's parsed.

@geoffkizer @stephentoub there are lots of places whose logic completely depends on RawValue nullability check. Simply stopping [resetting RawValue in ParseRawHeaderValues](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L763) failed +100 tests and even triggered Debug.Assert. I tried replacing those `RawValue == null` checks by introducing a new `bool IsParsed` property. Eventually by trials and errors, I managed to fix about 99 tests, but it required **making lots of dangerous changes to the existing parsing logic**. That version is still in this PR history. Please, take a look [at this commit](https://github.com/dotnet/runtime/pull/40713/commits/6c5bc7e6e09d2f58269e58406f87e0bfcb184f62) to see if that approach was better.

> Second, I think we can avoid the overhead of keeping the RawValue in sync with any changes the user makes to ParsedValue by simply clearing the RawValue if they make changes. We can then generate a new RawValue if/when they enumerate NonValidatedHeaders. This also fixes the update issue I mentioned above.

Conversion Raw -> Parsed -> Raw doesn't get us the original raw value in a common case. Could you please clarify how do you see it?

alnikola commented:
> Question: What happens if a user modifies an existing ParsedValue? E.g. update a single-valued header, or remove an item from a multi-valued header. Seems like we are not handling this case.

- Do you mean updating single-value header through `SetParsedValue`? If yes, it's already covered by [resetting OriginalRawValue to null](https://github.com/dotnet/runtime/blob/ca0e2cdb39c0da19a261b258e692ab4b95ea1185/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L387).
- I overlooked removal here, but propose to simply reset `OriginalRawValue` in this case, and later during non-validated enumeration convert `Parsed` to a string. It can return a value different from the original, but AFAIK the main goal of this PR is to implement enumeration of raw values arrived on the wire, so it should be an acceptable workaround. What do you think?

alnikola committed:
- Updating and removing parsed value is supported
- More unit tests

alnikola commented:
/azp run runtime-libraries outerloop

karelz commented:
Notes from team discussion (incl. @geoffkizer @scalablecory @alnikola and @dotnet/ncl):
- We agreed that given where we are in the product lifecycle, we need a safe change.
- Currently we know only about 1 safe change (this PR) -- it adds a field and has some perf impact. @alnikola believes that it does not affect the optimized code path where `TryAddValueWithoutValidation` is called and no strong-typed property is accessed -- hence no parsing happens. @geoffkizer will double check that.
- We are willing to pay the price of the extra field to have this API. We may choose to optimize it in future (6.0+).
- We will measure the impact of the new field anyway, just in case after the PR is merged on E2E scenarios. If it is bad, we have still some option to back out the change in RC1.
- @geoffkizer plans to experiment next couple of days to see if he can safely get the field back.

alnikola committed:
Fix ref

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,158 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+        public struct Enumerator : IEnumerator<KeyValuePair<string, HeaderStringValues>>
+        {
+            private readonly IEnumerator<KeyValuePair<HeaderDescriptor, object>> _headerStore;
```
We should avoid boxing here by using the actual struct type returned from Dictionary.GetEnumerator.

stephentoub replied:
(At which point the field can't be readonly, either.)

alnikola replied:
Done

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1357,158 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+        public struct Enumerator : IEnumerator<KeyValuePair<string, HeaderStringValues>>
+        {
+            private readonly IEnumerator<KeyValuePair<HeaderDescriptor, object>> _headerStore;
```
Calling this field `_headerStore` is confusing. Let's change it to something liek `_headerStoreEnumerator.

alnikola replied:
Done

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1366,166 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+            void IEnumerator.Reset()
+            {
+                _values?.Reset();
```
You don't actually need to implement Reset, you can just throw NotImplementedException. It's not used by foreach or LINQ or anything like that.

alnikola replied:
I know about that, but throwing NotImplementedException from IEnumerator.Reset is a bad choice for 3rd party library developers. If you writing some generic code working with IEnumerator, you'd expect it to fully implement contract. I personally ran into this issue in the past.

stephentoub replied:
> If you writing some generic code working with IEnumerator, you'd expect it to fully implement contract. I personally ran into this issue in the past.

No. The vast majority of Reset methods throw.  And it's perfectly acceptable to do so, even called out in the docs as such:
"The Reset method is provided for COM interoperability. It does not necessarily need to be implemented; instead, the implementer can simply throw a NotSupportedException."
No one can depend on an arbitrary Reset method working.

alnikola replied:
Ok, then. I will change it to throw NIE.

alnikola replied:
Done

geoffkizer commented:
> I overlooked removal here, but propose to simply reset OriginalRawValue in this case, and later during non-validated enumeration convert Parsed to a string. It can return a value different from the original, but AFAIK the main goal of this PR is to implement enumeration of raw values arrived on the wire, so it should be an acceptable workaround. What do you think?

I think this is fine. In general if the user modifies the value in any way (add/set/remove) then it's fine to discard the OriginalRawValue and generate a new one; the old one won't be accurate anyway.

geoffkizer commented on src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs:
```diff
@@ -1646,6 +1647,177 @@ public void GetEnumerator_UseExplicitInterfaceImplementation_EnumeratorReturnsNo
+        }
+
+        [Fact]
+        public void NonValidated_FirstHeaderWithOneValueSecondHeaderWithTwoValues_EnumeratorReturnsTwoHeaders()
```
Can we add a test that does something like this:

(1) Add raw header via TryAddWithoutValidation
(2) Retrieve it, causing it to be parsed
(3) Use NonValidatedHeaders and confirm that it's still returning the original raw value, unmodified

alnikola replied:
Added [NonValidated_AddRawHeaderValueThenRetrieveItAsParsed_EnumeratorReturnsOriginalRawValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1907) test.

geoffkizer commented on src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs:
```diff
@@ -1646,6 +1647,177 @@ public void GetEnumerator_UseExplicitInterfaceImplementation_EnumeratorReturnsNo
+        }
+
+        [Fact]
+        public void NonValidated_AddAndUpdateSingleValueHeader_EnumeratorReturnsLatestValue()
```
Can we add tests similar to this one and the next two, except that we start with a raw value instead of parsed values?

alnikola replied:
Done. Added tests
- [NonValidated_AddAndUpdateRawSingleValueHeader_EnumeratorReturnsLatestValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1823)
- [NonValidated_AddAndRemoveRawSingleValueHeader_EnumeratorReturnsNoValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1848)
- [NonValidated_AddAndRemoveRawValueToMultivalueHeader_EnumeratorReturnsLatestParsedValue](https://github.com/dotnet/runtime/blob/9c860a47d50cedbbe60f49ef891cd0b453bb0e9e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1880)

geoffkizer reviewed (changes requested):
Overall this is looking pretty good. I added some feedback above. 

I do think we need some more tests here. This is a somewhat scary change and the more tests we have, the better. I added comments about the tests I think we should add, but please also think about this holistically and add any other tests you think are useful.

stephentoub commented on src/libraries/System.Net.Http/ref/System.Net.Http.cs:
```diff
@@ -505,6 +522,20 @@ public abstract partial class HttpHeaders : System.Collections.Generic.IEnumerab
     }
+    public readonly partial struct HttpHeadersNonValidated : System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, System.Net.Http.Headers.HeaderStringValues>>, System.Collections.IEnumerable
+    {
+        public System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, System.Net.Http.Headers.HeaderStringValues>> GetEnumerator() { throw null; }
```
This needs to be `Enumerator`, not `System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, System.Net.Http.Headers.HeaderStringValues>>`

alnikola replied:
I can change it of course, but it was done the same way [as HttpHeaders.GetEnumerator](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/ref/System.Net.Http.cs#L515), so I thought it's intentional. No?

```C#
public System.Collections.Generic.IEnumerator<System.Collections.Generic.KeyValuePair<string, System.Collections.Generic.IEnumerable<string>>> GetEnumerator() { throw null; }
```

alnikola replied:
[HttpHeaders.GetEnumerator returns IEnumerator](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L308) because if `_headerStore` is empty it returns Array.Empty<>.Enumerator.

stephentoub replied:
The whole point of having the struct enumerator is to avoid the allocation.  This needs to change.  HttpHeaders.GetEnumerator uses Array.Empty's enumerator as an implementation detail _because_ the API was defined to return the interface.  If it had been defined to return a struct, it would have been implemented differently.

alnikola replied:
Ok. That's clear, then.

alnikola replied:
Done.

alnikola committed:
- IEnumerator field replaced with structs
- More tests added

alnikola committed:
AddRawHeaderValueThenRetrieveItAsParsed test fixed

alnikola commented:
@geoffkizer I addressed all the comments, could you please review and approve if everything is fine now?

alnikola commented:
/azp run runtime-libraries outerloop

alnikola commented:
Please, **merge the PR** after the final approval

alnikola commented:
There are several failed outerloop tests, but all of them are **unrelated** to this PR. Specifically,
- System.Text.RegularExpressions.Tests Work Item
- System.Security.Cryptography.X509Certificates.Tests.RevocationTests.DynamicRevocationTests.RevokeIntermediateWithExpiredRevocation
- System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2.Http2_PingKeepAlive

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<string>
```
What behavior do you expect from:
```C#
var e = default(Enumerator);
Console.WriteLine(e.MoveNext());
Console.WriteLine(e.Current);
```
?

alnikola replied:
Fixed.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+        }
+
+        public HeaderStringValues(IEnumerable<string> value)
+        {
```
```C#
if (value is null) throw new ArgumentNullException(nameof(value));
```
?

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+            _values = null;
+        }
+
+        public HeaderStringValues(IEnumerable<string> value)
```
Do we need these constructors to be public in this release?

stephentoub replied:
Also, shouldn't this be `values` rather than `value`?

alnikola replied:
They were approved [as public in API review](https://github.com/dotnet/runtime/issues/35126#issuecomment-670095983). Naming will be fixed.

geoffkizer replied:
Still, I think Stephen's point is that these don't need to be public, so we may as well keep them private for now.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+                return true;
+            }
+
+            public string Current { get; private set; }
```
You have this defined as `readonly` in the ref.  The get here needs to be made `readonly`.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
```
You have this defined as `readonly` in the ref.  The get here needs to be made `readonly`.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs:
```diff
@@ -1337,4 +1366,172 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
+
+            public void Dispose()
+            {
+                _headerStoreEnumerator.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                throw new NotImplementedException();
+            }
```
This has a mix of styles for one line methods.  They can all use the `=>` expression-bodied syntax.

geoffkizer commented:
It looks like there are still some places where parsed values are modified but original raw values are left unchanged. From a quick pass, these include:

AddParsedValue
CloneAndAddValue
TryParseAndAddValue

alnikola committed:
Enumerators fixed

alnikola committed:
- AddParsed covered by tests
- AddHeaderInfo fixed and covered by tests
- TryParseAndAddValue fixed and covered by tests

alnikola commented:
@geoffkizer I checked all places where `AddValue` adds Parsed or Invalid values, added there proper `OriginalRawValue` updates and covered everything with tests. Specifically,
- AddParsedValue - `OriginalRawValue` is already updated properly in [GetOrCreateHeaderInfo(descriptor, **parseRawValues: true**) call](https://github.com/dotnet/runtime/blob/0ee8de0cc529f3258fe402733c611ad51151bbac/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L370). Covered by tests.
- CloneAndAddValue. Fixed and covered by tests.
- TryParseAndAddValue.  Fixed and covered by tests.

Now it seems all places where header values get updated are fixed and covered by tests. Please, review and merge if it's fine now.
cc: @karelz

alnikola commented:
/azp run runtime-libraries outerloop

alnikola commented:
/azp run runtime-libraries outerloop

alnikola committed:
Small comment fix

geoffkizer commented:
> AddParsedValue - OriginalRawValue is already updated properly in GetOrCreateHeaderInfo(descriptor, parseRawValues: true) call. Covered by tests.

That's not the scenario I'm worried about.

What happens if the user does the following:

(1) Add raw header value using TryAddWithoutValidation
(2) Add parsed value via AddParsedValue

Now, if I understand the logic correctly, we'll end up with one OriginalRawValue from (1) and two parsed values. So NonValidatedEnumerator will only return one value, while the user sees two via the strongly typed API.

geoffkizer commented:
There's one other issue that just occurred to me that I don't think we are handling correctly.

Some strongly typed header values are mutable. For example, `MediaTypeHeaderValue`. See https://docs.microsoft.com/en-us/dotnet/api/system.net.http.headers.mediatypeheadervalue?view=netcore-3.1

That means the user can retrieve the strongly-typed header value (causing it to be parsed) and modify it without the HttpHeaders object itself knowing that it has been modified. Which means the parsed value and the original raw value will be out of sync.

I think the implication here is that anything that causes a header value to be parsed needs to discard the original raw value. Which is what the original code here does, actually, so that may end up simplifying the implementation here. But it also means that you'll only be able to access the original raw value if you literally never access the header value in any other way than accessing via NonValidatedHeaders. That's kinda weird, but I guess it's probably fine because it still serves the necessary purpose.

geoffkizer commented:
I think that also lets us get rid of OriginalRawValue now, because there's now no case in which we can hold on to the OriginalRawValue when we don't have the RawValue itself.

alnikola commented:
I see you point, but then non-validating enumeration behavior will get really confusing, because the value returned will depend on calls to strongly-typed **getters**. It seems very unnatural.

geoffkizer commented:
> It seems very unnatural.

I agree, it's very weird. But I'm not sure what else to do here. Suggestions?

alnikola commented:
> > AddParsedValue - OriginalRawValue is already updated properly in GetOrCreateHeaderInfo(descriptor, parseRawValues: true) call. Covered by tests.
> 
> That's not the scenario I'm worried about.
> 
> What happens if the user does the following:
> 
> (1) Add raw header value using TryAddWithoutValidation
> (2) Add parsed value via AddParsedValue
> 
> Now, if I understand the logic correctly, we'll end up with one OriginalRawValue from (1) and two parsed values. So NonValidatedEnumerator will only return one value, while the user sees two via the strongly typed API.

You are right, I overlooked the fact that value passed to AddParsedValue is not passed further down to GetOrCreateHeaderInfo, so OriginalRawValue is left unchanged, but I can easily fix it if this was the only concern.

alnikola commented:
> > It seems very unnatural.
> 
> I agree, it's very weird. But I'm not sure what else to do here. Suggestions?

It seems the main blocker is scenario when user updated parsed value. BUT, non-validating enumeration is supposed to return **raw** value added to the collection. Thus, it seems reasonable to say that all direct changed to **parsed** values are not taken into account because the don't change the original **raw** value.

alnikola commented:
@geoffkizer If you agree with the above API interpretation I can easily fix the last issue in AddParsedValue, cover it with test and we the could merge the PR. Frankly, I don't see other reasonable way to do it.

geoffkizer commented:
> It seems the main blocker is scenario when user updated parsed value. BUT, non-validating enumeration is supposed to return raw value added to the collection. Thus, it seems reasonable to say that all direct changed to parsed values are not taken into account because the don't change the original raw value.

Possibly. Right now it seems we are taking some changes to parsed values into account, but not others. Just ignoring all changes made to parsed values would at least be consistent. But it's also pretty weird itself, because it doesn't represent the overall state of the header collection in any meaningful way.

I think I'd prefer to just discard raw values when they are parsed.

geoffkizer commented:
> If you agree with the above API interpretation I can easily fix the last issue in AddParsedValue, cover it with test and we then could merge the PR.

It's more than just that one issue though, right? 

If we say what you suggest above:

> all direct changed to parsed values are not taken into account because the don't change the original raw value

Then we also need to change the handling in several other places, like SetParsedValue, to *not* remove the OriginalRawValue.

alnikola committed:
Enumerator tracks external changes to parsed values and returns reconstructed value in they are detected

alnikola commented:
@geoffkizer I added an tracking of external changes to parsed value (by storing a clone of the last known value) and fixed AddParsedValue which now resets OriginalRawValue the same way as SetParsedValue. Relevant tests are added as well.

It seems like PR now covers everything we need.

geoffkizer commented:
This is quite cool, thanks for doing this. I agree that this approach actually ends up being cleaner.

alnikola commented:
@geoffkizer So how do we proceed from here on? Will you approve it and then I merge?

geoffkizer commented:
Here's a thought I had on this after our discussion the other day. It was partially inspired by @stephentoub's comment that we shouldn't need to box values like ContentLength.

The way we manage parsed values is unnecessarily complex. It's designed to be super-general, but the reality is, there's a fixed number of parsed headers and we add to these very rarely. 

The current approach has essentially three layers of header processing:

(1) Strongly-typed header objects, meaning `HttpRequestHeaders`, `HttpResponseHeaders`, `HttpContentHeaders`. These are thin wrappers over (2). (Well, actually not that thin when dealing with multi-valued headers, because we end up creating a whole new `HttpHeaderValueCollection` on top of the underlying `List<T>` of parsed values in the HttpHeaders object.
(2) Parsed values at the HttpHeaders layer
(3) Raw values at the HttpHeaders layer

I think the system would be a lot simpler and more efficient if we reduced these to two layers. The strongly-typed header objects should be responsible for parsing, storing parsed values, tracking changes, and updating raw values when necessary. The HttpHeaders header storage should just deal with raw values. 

I think this would end up with a much better system. For example, simple values like Content-Length are actually stored as an int. Tracking changes is much simpler because we can know whether the value is single or multi-valued, whether it's immutable or not, etc and choose the appropriate change tracking model for each. 

Thoughts?

I can hack together an example of this, if it seems useful...

stephentoub commented:
@geoffkizer, @alnikola, is the plan to fix up this PR soon and keep going with this, or to start over with a different approach (in which case this should be closed)?

geoffkizer commented:
We cut this from 5.0 because we hit additional issues that were non-trivial to fix. Unfortunately, they are not documented above, but I can dig up old email if you're interested.

So there's more work here than just fixing up this PR. I think we should close this PR and revisit the entire issue.

alnikola commented:
Agree. Let's close this PR for now.
